data-processing/pdf/process_pdf.py

```diff
@@ -0,0 +1,236 @@
+import requests
+import logging
+from typing import List, Dict
+import sys
+from tempfile import TemporaryDirectory
+import subprocess
+import time
+
+import pdf.grobid_client
+from common.models import BoundingBox as BoundingBoxModel
+from common.models import init_database_connections
+from common.types import CitationData, CitationLocation, SerializableReference, BoundingBox, Matches, Match, Symbol
+from entities.symbols.types import SymbolData, SymbolWithId, SymbolId
+from entities.citations.utils import upload_citations, extract_ngrams, ngram_sim
+from entities.symbols.utils import upload_symbols
+
+class PdfStructureParser(object):
+    def __init__(self, pdf_hash, structure_map):
+        self.pdf_hash = pdf_hash
+        self.structure_map = structure_map
+        self.pages = structure_map['tokens']['pages']
+        self.page_sizes = dict((p['page']['pageNumber'], p['page']) for p in self.pages)
+        self.page_indices = dict((p['page']['pageNumber'], i) for i,p in enumerate(self.pages))
+        self.elements = structure_map['elements']['elementTypes']
+        self.references = None
+
+    def find_token(self, index):
+        for p in self.pages:
+            if (index < len(p['tokens'])):
+                return p['page'], p['tokens'][index]
+            else:
+                index -= len(p['tokens'])
+        return None
+
+
+    def get_text(self,spans):
+        text = []
+        for s in spans:
+            direct = s.get('dehyphenizedText')
+            if direct:
+                text += direct
+            else:
+                for i in range(s['left'], s['right']):
+                    page, token = self.find_token(i)
+                    text.append(token['text'])
+        return ' '.join(text)
+
+    def union(self, bbox1: BoundingBoxModel, bbox2: BoundingBoxModel):
+        if not bbox1:
+            return bbox2
+        x1 = min(bbox1.left,bbox2.left)
+        y1 = min(bbox1.top,bbox2.top)
+        x2 = max(bbox1.left+bbox1.width, bbox2.left+bbox2.width)
+        y2 = max(bbox1.top+bbox1.height, bbox2.top+bbox2.height)
+        return BoundingBoxModel(page = bbox1.page,
+                                left = x1, top = y1, width = x2-x1, height=y2-y1)
+
+    @staticmethod
+    def should_combine(bbox1: BoundingBoxModel, bbox2: BoundingBoxModel):
+        return bbox1.page == bbox2.page and (
+            abs(bbox1.top - bbox2.top) < 4 # Same y-coordinate
+        ) and (
+            abs(bbox2.left - bbox1.left - bbox1.width) < 15 # To the right
```

Comment:
Should this be comparing the box2 left to box1 right? Doesn't this equation mean that this would be considered overlapping, because bbox1 right would be less than 15 pixels away from bbox2 left?

```
------      ------
|box1|<-14->|box2|
------      ------
```

Reply:
There's no `.right` field on `BoundingBox`, so it's comparing to `bbox1.left+bbox1.width`.

Reply:
Got it! I misunderstood the intent of this code initially (I thought it was trying to merge overlapping bounding boxes, not adjacent ones). This looks good.